src/wasm-type.h

```diff
@@ -164,12 +190,161 @@ struct Signature {
   }
   bool operator!=(const Signature& other) const { return !(*this == other); }
   bool operator<(const Signature& other) const;
+  std::string toString() const;
+};
+
+struct Field {
+  Type type;
+  bool mutable_;
+  Field(Type type, bool mutable_ = false) : type(type), mutable_(mutable_) {}
+  bool operator==(const Field& other) const {
+    return type == other.type && mutable_ == other.mutable_;
+  }
+  bool operator!=(const Field& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+typedef std::vector<Field> FieldList;
+
+struct Struct {
+  FieldList fields;
+  Struct(const Struct& other) : fields(other.fields) {}
+  Struct(FieldList fields) : fields(fields) {}
+  bool operator==(const Struct& other) const { return fields == other.fields; }
+  bool operator!=(const Struct& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+struct Array {
+  Field element;
+  Array(const Array& other) : element(other.element) {}
+  Array(Field element) : element(element) {}
+  bool operator==(const Array& other) const { return element == other.element; }
+  bool operator!=(const Array& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+union TypeDef {
+  enum Kind { TupleKind, SignatureKind, StructKind, ArrayKind };
+
+  struct TupleDef {
+    Kind kind;
+    Tuple tuple;
+  } tupleDef;
+  struct SignatureDef {
+    Kind kind;
+    Signature signature;
+    bool nullable;
+  } signatureDef;
+  struct StructDef {
+    Kind kind;
+    Struct struct_;
+    bool nullable;
+  } structDef;
+  struct ArrayDef {
+    Kind kind;
+    Array array;
+    bool nullable;
+  } arrayDef;
```

Comment:
The difference is that an `Array` is a `DefType` conceptionally (as represented in the `type` section), with what's `nullable` being a constructed type of it (via `ArrayDef`). Some thoughts [here](https://github.com/WebAssembly/binaryen/pull/3012#discussion_r464087404).

Regarding the your first point, the sole reason for squeezing the `kind` into the defs (tried what you suggest earlier) was that it otherwise required a lot of boilerplate code upon construction of a `TypeDef`, like copy constructors for each kind, a switch within the `TypeDef` constructor that kept erroring on me with `new` placement not working as I wanted. Now I'm not sure if that's merely me not knowing how to do it properly, or if it's actually problematic to express. An alternative might be to derive all the `XYDef`s from a base `Def` struct that only has `kind`, but that'd still be just a convention. Hmm...

Reply:
Ugh, C++. I'll clone this branch locally to take a look at the problems you run into when you extract `kind` into a separate field.

If we look at `nullable X` as a constructed type of X, would it make sense to give `Nullable` its own union variant?

Reply:
Hmm, good question. Thinking about it more, we'll even want a union over all kinds allowed in the type section (currently keyed as `Signature`, maybe `union { Signature, Struct, Array }`), but that'd exclude `Tuple` and `Basic` (the latter as suggested elsewhere). Becomes either cyclic or redundant pretty quick, and then there'd also be `Nullable`. Oh my.

Reply:
Oh right, I haven't considered how to create a recursive type. I think it will be possible, but slightly more complicated, so let's just not think about that for now.

I don't think the type section is the right thing to think about here, but rather we should consider what types stack values and locals can have. For example, in MVP WebAssembly the type section contains signatures, but no stack value can have a signature type, so we don't bother representing signatures as a `Type`.

Reply:
Alright, let's postpone that. Regarding `Nullable`, what I've been wondering is whether we actually need to have nullable types other than `Struct` and `Array` (and perhaps RTT at some point). For instance, `null anyref ::= anyref` and `null externref ::= externref`, but I'm not 100% sure about `null funcref ::= funcref` (still a bit confused about `funcref` not being a subtype of `externref` now, but iiuc that was just about different representation of function and object references, so `null funcref ::= funcref` should hold?). If that's the case, I guess not introducing a recursive `Nullable` may make matters a little less complicated.

Reply:
Ok, so we could just have an `isNullable` predicate that always returns `true` for externref and funcref, and looks at a bool `nullable` that is a member of `Struct` and `Array`. (Yes, my understanding that `funcref` is nullable). That does seem simpler than adding a whole new `Nullable` variant.